plugin/pkg/bastion/validator/admission.go

```diff
@@ -0,0 +1,156 @@
+func (v *Bastion) AssignReadyFunc(f admission.ReadyFunc) {
+	v.readyFunc = f
+	v.SetReadyFunc(f)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (v *Bastion) SetInternalCoreClientset(c coreclientset.Interface) {
+	v.coreClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *Bastion) ValidateInitialization() error {
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	return nil
+}
+
+var _ admission.MutationInterface = &Bastion{}
+
+// Admit validates and if appropriate mutates the given bastion against the shoot that it references.
+func (v *Bastion) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Bastion
+	if a.GetKind().GroupKind() != operations.Kind("Bastion") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	// Convert object to Bastion
+	bastion, ok := a.GetObject().(*operations.Bastion)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert object to Bastion")
+	}
+
+	gk := schema.GroupKind{Group: operations.GroupName, Kind: "Bastion"}
+
+	// ensure shoot name is specified
+	shootPath := field.NewPath("spec", "shootRef", "name")
+	if bastion.Spec.ShootRef.Name == "" {
+		return apierrors.NewInvalid(gk, bastion.Name, field.ErrorList{field.Required(shootPath, "shoot is required")})
+	}
+
+	// ensure shoot exists
+	shootName := bastion.Spec.ShootRef.Name
+	shoot, err := v.coreClient.Core().Shoots(bastion.Namespace).Get(ctx, shootName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			fieldErr := field.Invalid(shootPath, shootName, fmt.Sprintf("shoot %s/%s not found", bastion.Namespace, shootName))
+			return apierrors.NewInvalid(gk, bastion.Name, field.ErrorList{fieldErr})
+		}
+
+		return apierrors.NewInternalError(fmt.Errorf("could not get shoot %s/%s: %v", bastion.Namespace, shootName, err))
+	}
+
+	// ensure shoot is already assigned to a seed
+	if shoot.Spec.SeedName == nil || len(*shoot.Spec.SeedName) == 0 {
+		fieldErr := field.Invalid(shootPath, shootName, "shoot is not yet assigned to a seed")
+		return apierrors.NewInvalid(gk, bastion.Name, field.ErrorList{fieldErr})
+	}
+
+	// update bastion
+	bastion.Spec.SeedName = shoot.Spec.SeedName
+	bastion.Spec.ProviderType = &shoot.Spec.Provider.Type
+
+	if userInfo := a.GetUserInfo(); userInfo != nil {
+		metav1.SetMetaDataAnnotation(&bastion.ObjectMeta, v1beta1constants.GardenCreatedBy, userInfo.GetName())
+	}
```

Comment:
This should only be done once for the initial creation - otherwise, nobody except the initial creator of the resource (including the to-be-introduced gardenlet controller) won't be able to change the resource:

```
k annotate bastion example-bastion gardener.cloud/operation=keepalive --overwrite --as=foo --as-group=system:masters
The Bastion "example-bastion" is invalid: metadata.annotations: Invalid value: "foo": field is immutable
```

Reply:
Fixed.

Reply:
good catch, however how is this code related to the annotations being immutable?

Reply:
@petersutter This code always sets the `created-by` annotation value to the user making the request. In `pkg/apis/operations/validation`, it is checked that this annotation never changes.

Reply:
ah ok got it